sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java

```diff
@@ -638,34 +652,132 @@ Mono<EncryptedBlob> encryptBlob(Flux<ByteBuffer> plainTextFlux) {
         }
     }
 
+    private Flux<ByteBuffer> encryptV2(Flux<ByteBuffer> plainTextFlux, SecretKey aesKey) {
+        Flux<ByteBuffer> encryptedTextFlux;
+        BufferStagingArea stagingArea =
+            new BufferStagingArea(GCM_ENCRYPTION_REGION_LENGTH, GCM_ENCRYPTION_REGION_LENGTH);
+
+        encryptedTextFlux =
+            UploadUtils.chunkSource(plainTextFlux,
+                    new com.azure.storage.common.ParallelTransferOptions()
+                        .setBlockSizeLong((long) GCM_ENCRYPTION_REGION_LENGTH))
+                .flatMapSequential(stagingArea::write)
+                .concatWith(Flux.defer(stagingArea::flush))
+                .index()
+                .flatMapSequential(tuple -> {
+                    Cipher gmcCipher;
+                    try {
+                        // We use the index as the nonce as a counter guarantees each nonce is used
+                        // only once with a given key.
+                        gmcCipher = generateGCMCipher(aesKey, tuple.getT1());
+                    } catch (NoSuchPaddingException | NoSuchAlgorithmException
+                        | InvalidAlgorithmParameterException | InvalidKeyException e) {
+                        throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+                    }
+
+                    // Expected size of each encryption region after calling doFinal. Last one may
+                    // be less, will never be more.
+                    ByteBuffer encryptedRegion = ByteBuffer.allocate(
+                        GCM_ENCRYPTION_REGION_LENGTH + TAG_LENGTH);
+
+                    // Each flux is at most 1 BufferAggregator of 4mb
+                    Flux<ByteBuffer> cipherTextWithTag = tuple.getT2()
+                        .asFlux()
+                        .map(buffer -> {
+                            // Write into the preallocated buffer and always return this buffer.
+                            try {
+                                gmcCipher.update(buffer, encryptedRegion);
+                            } catch (ShortBufferException e) {
+                                throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+                            }
+                            return encryptedRegion;
+                        })
+                        .then(Mono.fromCallable(() -> {
+                            // We have already written all the data to the cipher. Passing in a final
+                            // empty buffer allows us to force completion and return the filled buffer.
+                            gmcCipher.doFinal(ByteBuffer.allocate(0), encryptedRegion);
```

Comment:
Given that the `ByteBuffer` being allocated is always empty could we turn this into a constant as this pattern is used elsewhere and it'll stop us from creating objects on the heap that aren't need. This can be pushed to a follow-up change.

Reply:
Added to list